pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java

```diff
@@ -562,6 +570,25 @@ private void handleUpsert(ImmutableSegment immutableSegment) {
     }
   }
 
+  // Same as the SegmentLocks, but just for handleUpsert method to synchronize threads doing segment replacement.
+  private static class UpsertSegmentLocks {
```

Comment:
We can make the class abstract and create 2 static instance one for generic and one for upsert. (This way we can extend more later)

Reply:
But all members in this lock util class are static. So I assume no need to define an abstract base class in this case.

Reply:
i was thinking: 
```
public class SegmentLocks {
  private static final SegmentLocks GENERIC_LOCK = new SegmentLocks();
  private static final SegmentLocks UPSERT_LOCK = new SegmentLocks();

  private final Lock[] _locks = new Lock[NUM_LOCKS];

  private SegmentLocks() {
    for (int i = 0; i < NUM_LOCKS; i++) {
      _locks[i] = new ReentrantLock();
    }
  }

  private Lock getLock(String tableNameWithType, String segmentName) {
    return _locks[Math.abs((31 * tableNameWithType.hashCode() + segmentName.hashCode()) % NUM_LOCKS)];
  }

  public static Lock getSegmentLock(String tableNameWithType, String segmentName) {
    return GENERIC_LOCK.getLock(tableNameWithType, segmentName);
  }

  public static Lock getUpsertLock(String tableNameWithType, String segmentName) {
    return UPSERT_LOCK.getLock(tableNameWithType, segmentName);
  }
```

Reply:
Not bad

Reply:
this way we can extend more if we have to (just add one more static member). sorry for the "abstract" comment, i realized that it was a static class afterwards

Reply:
Your change is neat. But on a second thought, I kinda prefer to keep this upsert lock util class private to RealtimeDataManager for now, until we find other classes that also need to sync some upsert logic